### PR TITLE
[dhctl] Migrate to module config

### PR DESCRIPTION
**Pull request by name212**
## Description
move from config overrides to module config

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: <kebab-case of a module name> | <1st level dir in the repo>
type: fix | feature | chore
summary: <ONE-LINE of what effectively changes for a user>
impact: <what to expect for users, possibly MULTI-LINE>, required if impact_level is high ↓
impact_level: default | high | low
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by name212** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.25 6642907500 6642907500-1-con-1-25 /sys/deckhouse-oss/install:pr6311 cloud-user@158.160.51.42

**Comment by name212** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.25 6642907500 6642907500-1-con-1-25 /sys/deckhouse-oss/install:pr6311 cloud-user@158.160.51.42

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by name212** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.25 6649468200 6649468200-1-con-1-25 /sys/deckhouse-oss/install:pr6311 cloud-user@51.250.9.50

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by name212** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.25 6649881718 6649881718-1-con-1-25 /sys/deckhouse-oss/install:pr6311 cloud-user@158.160.63.106

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/eks'...

**Comment by name212** [comment]
/e2e/abort eks;WithoutNAT;containerd;1.25 6651143694 6651143694-1-con-1-25 /sys/deckhouse-oss/install:pr6311

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/eks'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/vsphere'...

**Comment by name212** [comment]
/e2e/abort eks;WithoutNAT;containerd;1.25 6651424942 6651424942-1-con-1-25 /sys/deckhouse-oss/install:pr6311

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/eks'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/azure'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/static'...

File: candi/openapi/init_configuration.yaml [hunk 1]
```diff
@@ -68,28 +68,43 @@ apiVersions:
             default: HTTPS
           releaseChannel:
             type: string
+            deprecated: true
             description: |
+              **Deprecated!** Use ModuleConfig 'deckhouse' for configure release channel.
```
Comment (mvasl) [inline candi/openapi/init_configuration.yaml:73]:
```suggestion
              **Deprecated!** Use ModuleConfig 'deckhouse' to configure release channel.
```

File: candi/openapi/init_configuration.yaml [hunk 2]
```diff
@@ -68,28 +68,43 @@ apiVersions:
...
+
               The [release channel](../modules/002-deckhouse/configuration.html#parameters-releasechannel) to use in the cluster.
             enum: [Alpha, Beta, EarlyAccess, Stable, RockSolid]
           devBranch:
             type: string
+            deprecated: true
             description: |
+              **Deprecated!** This field will be remove in the future and will be replaced on command-line argument.
```
Comment (mvasl) [inline candi/openapi/init_configuration.yaml:81]:
```suggestion
              **Deprecated!** This field will be removed in the future and replaced with command-line argument.
```

File: candi/openapi/init_configuration.yaml [hunk 3]
```diff
@@ -68,28 +68,43 @@ apiVersions:
...
+
               The parameter is used for development needs.
           bundle:
             type: string
+            deprecated: true
             description: |
+              **Deprecated!** Use ModuleConfig 'deckhouse' for configure bundle.
```
Comment (mvasl) [inline candi/openapi/init_configuration.yaml:88]:
```suggestion
              **Deprecated!** Use ModuleConfig 'deckhouse' to configure bundle.
```

File: candi/openapi/init_configuration.yaml [hunk 4]
```diff
@@ -68,28 +68,43 @@ apiVersions:
...
+
               The [Deckhouse bundle](../modules/002-deckhouse/configuration.html#parameters-bundle) to use in the cluster.
             enum: [Minimal, Managed, Default]
             default: Default
           logLevel:
             type: string
+            deprecated: true
             description: |
+              **Deprecated!** Use ModuleConfig 'deckhouse' for Deckhouse controller log level.
```
Comment (mvasl) [inline candi/openapi/init_configuration.yaml:97]:
```suggestion
              **Deprecated!** Use ModuleConfig 'deckhouse' to Deckhouse controller log level.
```

File: candi/openapi/init_configuration.yaml [hunk 5]
```diff
@@ -68,28 +68,43 @@ apiVersions:
...
+
               Deckhouse [logging Level](../modules/002-deckhouse/configuration.html#parameters-loglevel).
             enum: [Debug, Info, Error]
             default: Info
           configOverrides:
             type: object
+            deprecated: true
             description: |
+              **Deprecated!** Use ModuleConfig's resources in config.yaml for configure modules.
```
Comment (mvasl) [inline candi/openapi/init_configuration.yaml:106]:
```suggestion
              **Deprecated!** Use ModuleConfig's resources in config.yaml to configure modules.
```

File: dhctl/pkg/config/base.go [hunk 6]
```diff
@@ -182,9 +184,21 @@ func ParseConfigFromData(configData string) (*MetaConfig, error) {
...
+		err := yaml.Unmarshal(docData, moduleConfig)
+		if err == nil && moduleConfig.GetName() != "" {
+			_, err = schemaStore.Validate(&docData)
+			if err != nil {
+				return nil, fmt.Errorf("module config validation: %v\ndata: \n%s\n", err, numerateManifestLines(docData))
+			}
+			metaConfig.ModuleConfigs = append(metaConfig.ModuleConfigs, moduleConfig)
+			continue
+		} else {
+			index, err = schemaStore.Validate(&docData)
```
Comment (mvasl) [inline dhctl/pkg/config/base.go:198]: What is the point of that? ModuleConfigs are supposed to be valid yaml, aren't they?

File: dhctl/pkg/config/deckhouse_config.go [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+	tag := c.DevBranch
+	if forceVersionTag {
+		versionTag, foundValidTag := readVersionTagFromInstallerContainer()
+		if foundValidTag {
+			tag = versionTag
+		}
+	}
+
+	if tag == "" {
+		panic("Probably you use development image. please use devBranch")
```
Comment (mvasl) [inline dhctl/pkg/config/deckhouse_config.go:83]:
```suggestion
		panic("Probably you are using a development image. please use devBranch")
```

File: dhctl/pkg/config/deckhouse_config.go [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+		return nil, fmt.Errorf("marshal provider config: %v", err)
+	}
+
+	staticClusterConfig, err := metaConfig.StaticClusterConfigYAML()
+	if err != nil {
+		return nil, fmt.Errorf("marshal static config: %v", err)
+	}
+
+	bundle := "Default"
+	logLevel := "Info"
```
Comment (mvasl) [inline dhctl/pkg/config/deckhouse_config.go:131]: Are there any constants for these 2?

File: dhctl/pkg/config/deckhouse_config.go [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+
+	tag := string(rawFile)
+	if _, err = semver.NewVersion(tag); err != nil {
+		return "", false
+	}
+
+	return tag, true
+}
+
+func PrepareDeckhouseInstallConfig(metaConfig *MetaConfig) (*DeckhouseInstaller, error) {
```
Comment (mvasl) [inline dhctl/pkg/config/deckhouse_config.go:111]: This function is very long and has high cognitive complexity, there should be some way to split it

File: dhctl/pkg/config/deckhouse_config.go [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+kind: ModuleConfig
+metadata:
+  name: cni-flannel
+spec:
+  enabled: true
+---
+...
+`)
+		if len(metaConfig.ModuleConfigs) > 0 {
+			return nil, fmt.Errorf("Cannot use ModuleConfig's and configOverrides. Please use ModuleConfig's")
```
Comment (mvasl) [inline dhctl/pkg/config/deckhouse_config.go:191]:
```suggestion
			return nil, fmt.Errorf("Cannot use ModuleConfig's and configOverrides at the same time. Please use ModuleConfig's")
```

File: dhctl/pkg/config/deckhouse_config.go (hunk 9 above)
Comment (mvasl) [reply dhctl/pkg/config/deckhouse_config.go:111]: Like, at least the part where you handle deckhouse ModuleConfig does not belong here and can be extracted into a separate method or function.

File: dhctl/pkg/config/deckhouse_config.go [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+			"bundle":   bundle,
+			"logLevel": logLevel,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("Cannot create ModuleConfig deckhouse: %s", err)
+		}
+		metaConfig.ModuleConfigs = append(metaConfig.ModuleConfigs, deckhouseCm)
+	} else {
+		releaseChannelRaw, hasReleaseChannelKey := deckhouseCm.Spec.Settings["releaseChannel"]
+		if rc, ok := releaseChannelRaw.(string); hasReleaseChannelKey && ok {
```
Comment (mvasl) [inline dhctl/pkg/config/deckhouse_config.go:234]: Is it ok if there is releaseChannel but it's not a string, but some other junk?

File: dhctl/pkg/config/base.go [hunk 12]
```diff
@@ -182,9 +184,21 @@ func ParseConfigFromData(configData string) (*MetaConfig, error) {
 
 		docData := []byte(doc)
 
-		index, err := schemaStore.Validate(&docData)
-		if err != nil {
-			return nil, fmt.Errorf("config validation: %v\ndata: \n%s\n", err, numerateManifestLines(docData))
+		var index *SchemaIndex
+		moduleConfig := &ModuleConfig{}
+		err := yaml.Unmarshal(docData, moduleConfig)
+		if err == nil && moduleConfig.GetName() != "" {
```
Comment (name212) [inline dhctl/pkg/config/base.go:190]: use Kind

File: dhctl/pkg/config/deckhouse_config.go (hunk 8 above)
Comment (name212) [reply dhctl/pkg/config/deckhouse_config.go:131]: valid

File: dhctl/pkg/config/load.go [hunk 13]
```diff
@@ -77,6 +86,46 @@ func newSchemaStore(schemasDir []string) *SchemaStore {
...
 	}
 
+	entries, err := os.ReadDir(modulesDir)
+	if err != nil {
+		// autoconverger and state exporter do not contains module dir
+		log.WarnF("Modules dir not found")
+		return st
+	}
+
+	g := func(path string, moduleName string) {
```
Comment (name212) [inline dhctl/pkg/config/load.go:96]: rename

File: dhctl/pkg/config/load.go [hunk 14]
```diff
@@ -77,6 +86,46 @@ func newSchemaStore(schemasDir []string) *SchemaStore {
...
+		if !e.IsDir() {
+			continue
+		}
+		name := e.Name()
+		moduleName := strings.TrimLeft(name, "01234567890-")
+		p := path.Join(modulesDir, name, "openapi", "config-values.yaml")
+		g(p, moduleName)
+	}
+
+	g(path.Join(globalHooksModule, "openapi", "config-values.yaml"), "global")
```
Comment (name212) [inline dhctl/pkg/config/load.go:127]: panic here with comment

Comment (name212) [reply dhctl/pkg/config/load.go:127]: not panic log.Fatal

File: werf.yaml [hunk 15]
```diff
@@ -535,6 +539,47 @@ import:
...
   to: /usr/bin/dhctl
   after: setup
+- image: dev-prebuild
+  add: /deckhouse
+  to: /deckhouse
+  includePaths:
+  - modules/**/openapi/config-values.yaml
+  - global-hooks/openapi/config-values.yaml
+  excludePaths:
+  - modules/**/Chart.yaml
```
Comment (name212) [inline werf.yaml:549]: modules/**/

File: modules/040-node-manager/webhooks/validating/node_group [hunk 16]
```diff
@@ -56,18 +56,15 @@ kubernetes:
...
     executeHookOnEvent: []
     executeHookOnSynchronization: false
     keepFullObjectsInMemory: false
-    namespace:
-      nameSelector:
-        matchNames: ["d8-system"]
     nameSelector:
-      matchNames: ["${DECKHOUSE_CONFIG_MAP}"]
-    jqFilter: '.data.global // ""'
+      matchNames: ["deckhouse"]
```
Comment (name212) [inline modules/040-node-manager/webhooks/validating/node_group:66]: global

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/vsphere'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/static'...

**Comment by name212** [comment]
/e2e/abort static;Static;containerd;1.25 6657855349 6657855349-1-con-1-25 /sys/deckhouse-oss/install:pr6311 astra@192.168.199.71

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/static'...

**Comment by raabdullaev** [comment]
https://github.com/deckhouse/deckhouse/pull/6353

**Pull request merged**